lib/teleterm/clusters/cluster_auth.go

```diff
@@ -0,0 +1,258 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusters
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/client"
+	dbprofile "github.com/gravitational/teleport/lib/client/db"
+	"github.com/gravitational/teleport/lib/kube/kubeconfig"
+	web "github.com/gravitational/teleport/lib/web/ui"
+
+	"github.com/gravitational/trace"
+)
+
+// SyncAuthPreference fetches Teleport auth preferences and stores it in the cluster profile
+func (c *Cluster) SyncAuthPreference(ctx context.Context) (*web.WebConfigAuthSettings, error) {
+	_, err := c.clusterClient.Ping(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := c.clusterClient.SaveProfile(c.dir, false); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg, err := c.clusterClient.GetWebConfig(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &cfg.Auth, nil
+}
+
+// Logout deletes all cluster certificates
+func (c *Cluster) Logout(ctx context.Context) error {
+	// Delete db certs
+	for _, db := range c.status.Databases {
+		err := dbprofile.Delete(c.clusterClient, db)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Get the address of the active Kubernetes proxy to find AuthInfos,
+	// Clusters, and Contexts in kubeconfig.
+	clusterName, _ := c.clusterClient.KubeProxyHostPort()
+	if c.clusterClient.SiteName != "" {
+		clusterName = fmt.Sprintf("%v.%v", c.clusterClient.SiteName, clusterName)
+	}
+
+	// Remove cluster entries from kubeconfig
+	if err := kubeconfig.Remove("", clusterName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Remove keys for this user from disk and running agent.
+	if err := c.clusterClient.Logout(); !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// LocalLogin processes local logins for this cluster
+func (c *Cluster) LocalLogin(ctx context.Context, user, password, otpToken string) error {
+	pingResp, err := c.clusterClient.Ping(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// TODO(alex-kovoy): SiteName needs to be reset if trying to login to a cluster with
+	// existing profile for the first time (investigate why)
+	c.clusterClient.SiteName = ""
+
+	switch pingResp.Auth.SecondFactor {
+	case constants.SecondFactorOff, constants.SecondFactorOTP:
+		err := c.localLogin(ctx, user, password, otpToken)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	case constants.SecondFactorU2F, constants.SecondFactorWebauthn, constants.SecondFactorOn, constants.SecondFactorOptional:
```

Comment:
Yep, you shouldn't get "u2f" from the backend on the master version of Teleport. #10375 has some context about the U2F cleanup, if anyone is curious.

This should make things a bit simpler for Teleterm.

Reply:
According to the issue linked by @codingllama, aliasing u2f to webauthn didn't land in Teleport 9, so I'll leave `constants.SecondFactorU2F` here for now. It'll be trivial to remove this constant later since we don't do anything special for u2f in `lib/teleterm`.

The frontend code has a little bit of extra logic for u2f, but that's just like one or two functions.